packages/language/src/language-server/completion/completion-generator.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ *
+ */
+
+import { CompletionItemKind } from "vscode-languageserver-types";
+import { Statement, SyntaxKind, SyntaxNode } from "../../syntax-tree/ast";
+import { CompilationUnit } from "../../workspace/compilation-unit";
+import { CompletionItem } from "../types";
+import { FollowElement, FollowKind } from "./follow-elements";
+import { getQualifiedName } from "../../linking/resolver";
+
+export interface SimpleCompletionItem extends Omit<CompletionItem, "edit"> {
+  text: string;
+}
+
+export function generateCompletionItems(
+  unit: CompilationUnit,
+  context: SyntaxNode | undefined,
+  followElement: FollowElement,
+): SimpleCompletionItem[] {
+  const items: SimpleCompletionItem[] = [];
+  if (followElement.kind === FollowKind.CstNode) {
+    // TODO: implement completion for keywords
+    return [];
+  }
+  if (followElement.kind === FollowKind.QualifiedReference) {
+    context = followElement.previous;
+  } else if (!context) {
+    return [];
+  }
+  const scopeCache = isPreprocessorNode(context, unit)
+    ? unit.scopeCaches.preprocessor
+    : unit.scopeCaches.regular;
+  const scope = scopeCache.get(context);
+  if (!scope) {
+    return [];
+  }
+  if (followElement.kind === FollowKind.LocalReference) {
+    const symbols = scope.allDistinctSymbols([]);
+    for (const symbol of symbols) {
+      if (symbol.name) {
+        items.push({
+          label: symbol.name,
+          kind: getCompletionKind(symbol.node),
+          text: symbol.name,
+        });
+      }
+    }
+  } else if (followElement.kind === FollowKind.QualifiedReference) {
+    const ref = followElement.previous.element?.ref;
+    if (!ref) {
+      return [];
+    }
+    const qualifiedName = getQualifiedName(ref);
+    const symbols = scope.allDistinctSymbols(qualifiedName);
+    for (const symbol of symbols) {
+      if (symbol.name) {
+        items.push({
+          label: symbol.name,
+          kind: getCompletionKind(symbol.node),
+          text: symbol.name,
+        });
+      }
+    }
+  }
+  return items;
+}
+
+function getCompletionKind(node: SyntaxNode): CompletionItemKind {
+  switch (node.kind) {
+    case SyntaxKind.LabelPrefix: {
+      const statement = node.container as Statement;
+      const value = statement.value;
+      if (value && value.kind === SyntaxKind.ProcedureStatement) {
+        return CompletionItemKind.Function;
+      }
+      return CompletionItemKind.Variable;
+    }
+  }
+  return CompletionItemKind.Variable;
+}
+
+function isPreprocessorNode(node: SyntaxNode, unit: CompilationUnit): boolean {
+  let parent = node;
```

Comment:
I would have thought we would have a faster way to resolve the node type than this, like an additional prop or link to the root. Our container chains won't be so deep here, but if they are we may want to revisit this later.

Reply:
Right now we don't. I don't want to add a prop to the preprocessor types, since they are just "normal" AST nodes - but sourced from a different parser. I also think it should be generally fine, since it will always be fast enough and we perform this check only once per completion request.

Reply:
That was my thought as well, if it were in a hot path sure, but it's just one-of completions. Should be fine.